# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [6]:
import pandas as pd
import patsy
import numpy as np
import statsmodels.formula.api as smf
import statsmodels.api as sm
import patsy


In [7]:
df = pd.read_csv('previsao_de_renda.csv')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  


2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  


3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor. 
    

In [10]:

y, x = patsy.dmatrices( 'np.log(renda) ~ C(tipo_renda) + C(educacao) + C(estado_civil) + C(tipo_residencia)+ qtd_filhos + posse_de_veiculo + posse_de_imovel+ idade + tempo_emprego + qt_pessoas_residencia', data=df, return_type='dataframe')

x

Intercept  C(tipo_renda)[T.Bolsista]  C(tipo_renda)[T.Empresário]  \
0            1.0                        0.0                          1.0   
1            1.0                        0.0                          0.0   
2            1.0                        0.0                          1.0   
3            1.0                        0.0                          0.0   
4            1.0                        0.0                          0.0   
...          ...                        ...                          ...   
14994        1.0                        0.0                          1.0   
14995        1.0                        0.0                          1.0   
14997        1.0                        0.0                          0.0   
14998        1.0                        0.0                          1.0   
14999        1.0                        0.0                          0.0   

       C(tipo_renda)[T.Pensionista]  C(tipo_renda)[T.Servidor público]  \
0                               0.0                                0.0   
1                               0.0                                0.0   
2                               0.0                                0.0   
3                               0.0                                1.0   
4                               0.0                                0.0   
...                             ...                                ...   
14994                           0.0                                0.0   
14995                           0.0                                0.0   
14997                           0.0                                0.0   
14998                           0.0                                0.0   
14999                           0.0                                1.0   

       C(educacao)[T.Pós graduação]  C(educacao)[T.Secundário]  \
0                               0.0                        1.0   
1                               0.0                        0.0   
2                               0.0                        0.0   
3                               0.0                        0.0   
4                               0.0                        1.0   
...                             ...                        ...   
14994                           0.0                        0.0   
14995                           0.0                        1.0   
14997                           0.0                        0.0   
14998                           0.0                        0.0   
14999                           0.0                        1.0   

       C(educacao)[T.Superior completo]  C(educacao)[T.Superior incompleto]  \
0                                   0.0                                 0.0   
1                                   1.0                                 0.0   
2                                   1.0                                 0.0   
3                                   1.0                                 0.0   
4                                   0.0                                 0.0   
...                                 ...                                 ...   
14994                               0.0                                 1.0   
14995                               0.0                                 0.0   
14997                               1.0                                 0.0   
14998                               1.0                                 0.0   
14999                               0.0                                 0.0   

       C(estado_civil)[T.Separado]  ...  C(tipo_residencia)[T.Com os pais]  \
0                              0.0  ...                                0.0   
1                              0.0  ...                                0.0   
2                              0.0  ...                                0.0   
3                              0.0  ...                                0.0   
4                              0.0  ...                                0.

In [15]:
sm.OLS(y,x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.206
Method:                 Least Squares   F-statistic:                     141.6
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        16:25:15   Log-Likelihood:                -14868.
No. Observations:               12427   AIC:                         2.978e+04
Df Residuals:                   12403   BIC:                         2.996e+04
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               7.2927      0.261     27.981      0.000       6.782       7.804
C(tipo_renda)[T.Bolsista]               0.0855      0.268      0.319      0.750      -0.439       0.610
C(tipo_renda)[T.Empresário]             0.1130      0.017      6.821      0.000       0.080       0.145
C(tipo_renda)[T.Pensionista]           -0.2609      0.268     -0.974      0.330      -0.786       0.264
C(tipo_renda)[T.Servidor público]       0.0108      0.025      0.437      0.662      -0.038       0.059
C(educacao)[T.Pós graduação]           -0.1277      0.176     -0.724      0.469      -0.473       0.218
C(educacao)[T.Secundário]              -0.0910      0.080     -1.138      0.255      -0.248       0.066
C(educacao)[T.Superior completo]       -0.0276      0.080     -0.345      0.730      -0.185       0.130
C(educacao)[T.Superior incompleto]     -0.1239      0.086     -1.435      0.151      -0.293       0.045
C(estado_civil)[T.Separado]             0.1951      0.123      1.580      0.114      -0.047       0.437
C(estado_civil)[T.Solteiro]             0.1875      0.121      1.551      0.121      -0.050       0.425
C(estado_civil)[T.União]               -0.0906      0.028     -3.245      0.001      -0.145      -0.036
C(estado_civil)[T.Viúvo]                0.1049      0.128      0.818      0.413      -0.146       0.356
C(tipo_residencia)[T.Casa]             -0.1896      0.060     -3.157      0.002      -0.307      -0.072
C(tipo_residencia)[T.Com os pais]      -0.1378      0.067     -2.058      0.040      -0.269      -0.007
C(tipo_residencia)[T.Comunitário]      -0.3174      0.117     -2.715      0.007      -0.547      -0.088
C(tipo_residencia)[T.Estúdio]           0.1271      0.110      1.153      0.249      -0.089       0.343
C(tipo_residencia)[T.Governamental]    -0.2375      0.073     -3.246      0.001      -0.381      -0.094
posse_de_veiculo[T.True]                0.2815      0.015     18.875      0.000       0.252       0.311
posse_de_imovel[T.True]                 0.0572      0.016      3.688      0.000       0.027       0.088
qtd_filhos                             -0.2082      0.120     -1.737      0.082      -0.443       0.027
idade                                   0.0027      0.001      3.023      0.003       0.001       0.004
tempo_emprego                           0.0571      0.001     49.683      0.000       0.055       0.059
qt_pessoas_residencia                   0.2440      0.120      2.040      0.041       0.010       0.479
==============================================================================
Omnibus:                       24.399   Durbin-Watson:                   2.034
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.482
Skew:                           0.108 

Agora colocando as variaveis mais frequentes como casela de referencia.

In [18]:
x.sum(axis=0)

Intercept                               12427.000000
C(tipo_renda)[T.Bolsista]                   9.000000
C(tipo_renda)[T.Empresário]              3508.000000
C(tipo_renda)[T.Pensionista]                9.000000
C(tipo_renda)[T.Servidor público]        1268.000000
C(educacao)[T.Pós graduação]               26.000000
C(educacao)[T.Secundário]                7045.000000
C(educacao)[T.Superior completo]         4695.000000
C(educacao)[T.Superior incompleto]        558.000000
C(estado_civil)[T.Separado]               739.000000
C(estado_civil)[T.Solteiro]              1543.000000
C(estado_civil)[T.União]                  924.000000
C(estado_civil)[T.Viúvo]                  324.000000
C(tipo_residencia)[T.Casa]              11071.000000
C(tipo_residencia)[T.Com os pais]         674.000000
C(tipo_residencia)[T.Comunitário]          64.000000
C(tipo_residencia)[T.Estúdio]              75.000000
C(tipo_residencia)[T.Governamental]       360.000000
posse_de_veiculo[T.True]                 5267.

As categorias mais frequentes das variaveis qualitativas sao: 'C(tipo_renda)[T.Empresário]' , 'C(educacao)[T.Secundário]', 'C(estado_civil)[T.Solteiro]', 'C(tipo_residencia)[T.Casa]'.

In [81]:

y, x = patsy.dmatrices('np.log(renda) ~ C(tipo_renda, Treatment("Empresário")) + C(educacao, Treatment("Secundário")) + C(estado_civil, Treatment("Solteiro")) + C(tipo_residencia, Treatment("Casa")) + qtd_filhos + posse_de_veiculo + posse_de_imovel+ idade + tempo_emprego + qt_pessoas_residencia', data=df, return_type='dataframe')

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.206
Method:                 Least Squares   F-statistic:                     141.6
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:08:22   Log-Likelihood:                -14868.
No. Observations:               12427   AIC:                         2.978e+04
Df Residuals:                   12403   BIC:                         2.996e+04
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       7.3127      0.128     57.276      0.000       7.062       7.563
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]          -0.1130      0.017     -6.821      0.000      -0.145      -0.080
C(tipo_renda, Treatment("Empresário"))[T.Bolsista]             -0.0275      0.268     -0.103      0.918      -0.553       0.498
C(tipo_renda, Treatment("Empresário"))[T.Pensionista]          -0.3738      0.268     -1.395      0.163      -0.899       0.151
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]     -0.1022      0.027     -3.824      0.000      -0.155      -0.050
C(educacao, Treatment("Secundário"))[T.Primário]                0.0910      0.080      1.138      0.255      -0.066       0.248
C(educacao, Treatment("Secundário"))[T.Pós graduação]          -0.0367      0.158     -0.233      0.816      -0.346       0.273
C(educacao, Treatment("Secundário"))[T.Superior completo]       0.0633      0.015      4.108      0.000       0.033       0.094
C(educacao, Treatment("Secundário"))[T.Superior incompleto]    -0.0329      0.036     -0.919      0.358      -0.103       0.037
C(estado_civil, Treatment("Solteiro"))[T.Casado]               -0.1875      0.121     -1.551      0.121      -0.425       0.050
C(estado_civil, Treatment("Solteiro"))[T.Separado]              0.0076      0.036      0.210      0.834      -0.064       0.079
C(estado_civil, Treatment("Solteiro"))[T.União]                -0.2781      0.123     -2.256      0.024      -0.520      -0.036
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]                -0.0826      0.050     -1.653      0.098      -0.180       0.015
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]                0.1896      0.060      3.157      0.002       0.072       0.307
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]            0.0518      0.033      1.584      0.113      -0.012       0.116
C(tipo_residencia, Treatment("Casa"))[T.Comunitário]           -0.1278      0.101     -1.265      0.206      -0.326       0.070
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]                0.3166      0.093      3.402      0.001       0.134       0.499
C(tipo_residencia, Treatment("Casa"))[T.Governamental]         -0.0479      0.043     -1.106      0.269      -0.133       0.037
posse_de_veiculo[T.True]                                        0.2815      0.015     18.875      0.000       0.252       0.311
posse_de_imovel[T.True]                                         0.0572      0.016      3.688      0.000       0.027       0.088
qtd_filhos                                                     -0.2082      0.120     -1.737      0.082      -0.443       0.027
idade      

Removendo a variavel C(tipo_renda, Treatment("Empresário"))[T.Bolsista] , a qual possui p = 0.918.

In [84]:
x = x.drop('C(tipo_renda, Treatment("Empresário"))[T.Bolsista]', axis=1)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.207
Method:                 Least Squares   F-statistic:                     148.0
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:08:23   Log-Likelihood:                -14868.
No. Observations:               12427   AIC:                         2.978e+04
Df Residuals:                   12404   BIC:                         2.995e+04
Df Model:                          22                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       7.3128      0.128     57.280      0.000       7.063       7.563
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]          -0.1129      0.017     -6.823      0.000      -0.145      -0.080
C(tipo_renda, Treatment("Empresário"))[T.Pensionista]          -0.3738      0.268     -1.395      0.163      -0.899       0.151
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]     -0.1021      0.027     -3.823      0.000      -0.154      -0.050
C(educacao, Treatment("Secundário"))[T.Primário]                0.0909      0.080      1.138      0.255      -0.066       0.248
C(educacao, Treatment("Secundário"))[T.Pós graduação]          -0.0367      0.158     -0.233      0.816      -0.346       0.273
C(educacao, Treatment("Secundário"))[T.Superior completo]       0.0632      0.015      4.107      0.000       0.033       0.093
C(educacao, Treatment("Secundário"))[T.Superior incompleto]    -0.0329      0.036     -0.919      0.358      -0.103       0.037
C(estado_civil, Treatment("Solteiro"))[T.Casado]               -0.1875      0.121     -1.551      0.121      -0.425       0.049
C(estado_civil, Treatment("Solteiro"))[T.Separado]              0.0076      0.036      0.210      0.834      -0.064       0.079
C(estado_civil, Treatment("Solteiro"))[T.União]                -0.2781      0.123     -2.256      0.024      -0.520      -0.036
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]                -0.0825      0.050     -1.653      0.098      -0.180       0.015
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]                0.1896      0.060      3.157      0.002       0.072       0.307
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]            0.0517      0.033      1.584      0.113      -0.012       0.116
C(tipo_residencia, Treatment("Casa"))[T.Comunitário]           -0.1278      0.101     -1.265      0.206      -0.326       0.070
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]                0.3167      0.093      3.403      0.001       0.134       0.499
C(tipo_residencia, Treatment("Casa"))[T.Governamental]         -0.0479      0.043     -1.105      0.269      -0.133       0.037
posse_de_veiculo[T.True]                                        0.2815      0.015     18.884      0.000       0.252       0.311
posse_de_imovel[T.True]                                         0.0572      0.016      3.687      0.000       0.027       0.088
qtd_filhos                                                     -0.2082      0.120     -1.737      0.082      -0.443       0.027
idade                                                           0.0027      0.001      3.021      0.003       0.001       0.004
tempo_empre

Apos ser removido a categoria T.Bolsista o R ajustado do modelo aumentou de 0.206 para 0.207 mostrando que mesmo que muito pequena, a remoção desta variavel foi benéfica, fora essas mudanças nao houve uma mudança significativa nos parâmetros.

A com maior p agora é a variavel C(estado_civil, Treatment("Solteiro"))[T.Separado]

In [88]:
x = x.drop('C(estado_civil, Treatment("Solteiro"))[T.Separado]', axis=1)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.207
Method:                 Least Squares   F-statistic:                     155.1
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:08:27   Log-Likelihood:                -14868.
No. Observations:               12427   AIC:                         2.978e+04
Df Residuals:                   12405   BIC:                         2.994e+04
Df Model:                          21                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       7.3150      0.127     57.493      0.000       7.066       7.564
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]          -0.1128      0.017     -6.821      0.000      -0.145      -0.080
C(tipo_renda, Treatment("Empresário"))[T.Pensionista]          -0.3757      0.268     -1.403      0.161      -0.901       0.149
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]     -0.1022      0.027     -3.827      0.000      -0.155      -0.050
C(educacao, Treatment("Secundário"))[T.Primário]                0.0912      0.080      1.142      0.253      -0.065       0.248
C(educacao, Treatment("Secundário"))[T.Pós graduação]          -0.0374      0.158     -0.237      0.813      -0.347       0.272
C(educacao, Treatment("Secundário"))[T.Superior completo]       0.0633      0.015      4.109      0.000       0.033       0.093
C(educacao, Treatment("Secundário"))[T.Superior incompleto]    -0.0332      0.036     -0.926      0.354      -0.103       0.037
C(estado_civil, Treatment("Solteiro"))[T.Casado]               -0.1895      0.121     -1.573      0.116      -0.426       0.047
C(estado_civil, Treatment("Solteiro"))[T.União]                -0.2801      0.123     -2.278      0.023      -0.521      -0.039
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]                -0.0852      0.048     -1.765      0.078      -0.180       0.009
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]                0.1891      0.060      3.151      0.002       0.071       0.307
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]            0.0518      0.033      1.584      0.113      -0.012       0.116
C(tipo_residencia, Treatment("Casa"))[T.Comunitário]           -0.1281      0.101     -1.268      0.205      -0.326       0.070
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]                0.3166      0.093      3.402      0.001       0.134       0.499
C(tipo_residencia, Treatment("Casa"))[T.Governamental]         -0.0481      0.043     -1.109      0.268      -0.133       0.037
posse_de_veiculo[T.True]                                        0.2815      0.015     18.886      0.000       0.252       0.311
posse_de_imovel[T.True]                                         0.0570      0.015      3.681      0.000       0.027       0.087
qtd_filhos                                                     -0.2075      0.120     -1.732      0.083      -0.442       0.027
idade                                                           0.0027      0.001      3.070      0.002       0.001       0.004
tempo_emprego                                                   0.0571      0.001     49.692      0.000       0.055       0.059
qt_pessoas_

A com maior p agora é a variavel C(educacao, Treatment("Secundário"))[T.Pós graduação]

In [91]:
x = x.drop('C(educacao, Treatment("Secundário"))[T.Pós graduação]', axis=1)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.207
Method:                 Least Squares   F-statistic:                     162.8
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:08:29   Log-Likelihood:                -14868.
No. Observations:               12427   AIC:                         2.978e+04
Df Residuals:                   12406   BIC:                         2.993e+04
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       7.3148      0.127     57.494      0.000       7.065       7.564
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]          -0.1127      0.017     -6.817      0.000      -0.145      -0.080
C(tipo_renda, Treatment("Empresário"))[T.Pensionista]          -0.3755      0.268     -1.402      0.161      -0.900       0.149
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]     -0.1021      0.027     -3.824      0.000      -0.154      -0.050
C(educacao, Treatment("Secundário"))[T.Primário]                0.0914      0.080      1.144      0.253      -0.065       0.248
C(educacao, Treatment("Secundário"))[T.Superior completo]       0.0634      0.015      4.122      0.000       0.033       0.094
C(educacao, Treatment("Secundário"))[T.Superior incompleto]    -0.0330      0.036     -0.922      0.356      -0.103       0.037
C(estado_civil, Treatment("Solteiro"))[T.Casado]               -0.1895      0.121     -1.572      0.116      -0.426       0.047
C(estado_civil, Treatment("Solteiro"))[T.União]                -0.2800      0.123     -2.277      0.023      -0.521      -0.039
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]                -0.0851      0.048     -1.763      0.078      -0.180       0.010
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]                0.1892      0.060      3.153      0.002       0.072       0.307
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]            0.0518      0.033      1.586      0.113      -0.012       0.116
C(tipo_residencia, Treatment("Casa"))[T.Comunitário]           -0.1281      0.101     -1.268      0.205      -0.326       0.070
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]                0.3167      0.093      3.403      0.001       0.134       0.499
C(tipo_residencia, Treatment("Casa"))[T.Governamental]         -0.0480      0.043     -1.107      0.268      -0.133       0.037
posse_de_veiculo[T.True]                                        0.2814      0.015     18.886      0.000       0.252       0.311
posse_de_imovel[T.True]                                         0.0569      0.015      3.676      0.000       0.027       0.087
qtd_filhos                                                     -0.2075      0.120     -1.732      0.083      -0.442       0.027
idade                                                           0.0027      0.001      3.069      0.002       0.001       0.004
tempo_emprego                                                   0.0571      0.001     49.714      0.000       0.055       0.059
qt_pessoas_residencia                                           0.2435      0.120      2.036      0.042       0.009       0.478
===========

In [93]:
x = x.drop('C(educacao, Treatment("Secundário"))[T.Superior incompleto]', axis=1)

sm.OLS(y, x).fit().summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.207
Method:                 Least Squares   F-statistic:                     171.4
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:08:30   Log-Likelihood:                -14868.
No. Observations:               12427   AIC:                         2.978e+04
Df Residuals:                   12407   BIC:                         2.992e+04
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
==============================================================================================================================
                                                                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      7.3069      0.127     57.562      0.000       7.058       7.556
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]         -0.1123      0.017     -6.792      0.000      -0.145      -0.080
C(tipo_renda, Treatment("Empresário"))[T.Pensionista]         -0.3742      0.268     -1.398      0.162      -0.899       0.151
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]    -0.1016      0.027     -3.806      0.000      -0.154      -0.049
C(educacao, Treatment("Secundário"))[T.Primário]               0.0942      0.080      1.180      0.238      -0.062       0.251
C(educacao, Treatment("Secundário"))[T.Superior completo]      0.0661      0.015      4.372      0.000       0.036       0.096
C(estado_civil, Treatment("Solteiro"))[T.Casado]              -0.1892      0.121     -1.570      0.116      -0.425       0.047
C(estado_civil, Treatment("Solteiro"))[T.União]               -0.2797      0.123     -2.275      0.023      -0.521      -0.039
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]               -0.0846      0.048     -1.753      0.080      -0.179       0.010
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]               0.1880      0.060      3.135      0.002       0.070       0.306
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]           0.0510      0.033      1.561      0.119      -0.013       0.115
C(tipo_residencia, Treatment("Casa"))[T.Comunitário]          -0.1305      0.101     -1.292      0.196      -0.328       0.067
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]               0.3173      0.093      3.409      0.001       0.135       0.500
C(tipo_residencia, Treatment("Casa"))[T.Governamental]        -0.0473      0.043     -1.091      0.275      -0.132       0.038
posse_de_veiculo[T.True]                                       0.2814      0.015     18.887      0.000       0.252       0.311
posse_de_imovel[T.True]                                        0.0572      0.015      3.696      0.000       0.027       0.088
qtd_filhos                                                    -0.2072      0.120     -1.730      0.084      -0.442       0.028
idade                                                          0.0028      0.001      3.237      0.001       0.001       0.005
tempo_emprego                                                  0.0571      0.001     49.706      0.000       0.055       0.059
qt_pessoas_residencia                                          0.2434      0.120      2.035      0.042       0.009       0.478
==============================================================================
Omnibus:                       24.635   Durbin-Watson:                   2.033
Prob

In [95]:
x = x.drop('C(tipo_residencia, Treatment("Casa"))[T.Governamental]', axis=1)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.207
Method:                 Least Squares   F-statistic:                     180.8
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:08:44   Log-Likelihood:                -14869.
No. Observations:               12427   AIC:                         2.978e+04
Df Residuals:                   12408   BIC:                         2.992e+04
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
==============================================================================================================================
                                                                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      7.3072      0.127     57.563      0.000       7.058       7.556
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]         -0.1122      0.017     -6.788      0.000      -0.145      -0.080
C(tipo_renda, Treatment("Empresário"))[T.Pensionista]         -0.3727      0.268     -1.392      0.164      -0.898       0.152
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]    -0.1013      0.027     -3.797      0.000      -0.154      -0.049
C(educacao, Treatment("Secundário"))[T.Primário]               0.0922      0.080      1.156      0.248      -0.064       0.249
C(educacao, Treatment("Secundário"))[T.Superior completo]      0.0665      0.015      4.401      0.000       0.037       0.096
C(estado_civil, Treatment("Solteiro"))[T.Casado]              -0.1872      0.121     -1.553      0.120      -0.423       0.049
C(estado_civil, Treatment("Solteiro"))[T.União]               -0.2769      0.123     -2.253      0.024      -0.518      -0.036
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]               -0.0847      0.048     -1.755      0.079      -0.179       0.010
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]               0.1897      0.060      3.164      0.002       0.072       0.307
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]           0.0529      0.033      1.623      0.105      -0.011       0.117
C(tipo_residencia, Treatment("Casa"))[T.Comunitário]          -0.1287      0.101     -1.274      0.203      -0.327       0.069
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]               0.3189      0.093      3.428      0.001       0.137       0.501
posse_de_veiculo[T.True]                                       0.2814      0.015     18.891      0.000       0.252       0.311
posse_de_imovel[T.True]                                        0.0594      0.015      3.868      0.000       0.029       0.090
qtd_filhos                                                    -0.2048      0.120     -1.710      0.087      -0.440       0.030
idade                                                          0.0028      0.001      3.205      0.001       0.001       0.004
tempo_emprego                                                  0.0571      0.001     49.707      0.000       0.055       0.059
qt_pessoas_residencia                                          0.2412      0.120      2.017      0.044       0.007       0.476
==============================================================================
Omnibus:                       24.709   Durbin-Watson:                   2.033
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.798
Skew:                           0.109   Prob(JB):   

In [97]:
x = x.drop('C(educacao, Treatment("Secundário"))[T.Primário]', axis=1)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.207
Method:                 Least Squares   F-statistic:                     191.4
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:09:39   Log-Likelihood:                -14869.
No. Observations:               12427   AIC:                         2.977e+04
Df Residuals:                   12409   BIC:                         2.991e+04
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
==============================================================================================================================
                                                                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      7.3092      0.127     57.583      0.000       7.060       7.558
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]         -0.1118      0.017     -6.765      0.000      -0.144      -0.079
C(tipo_renda, Treatment("Empresário"))[T.Pensionista]         -0.3724      0.268     -1.391      0.164      -0.897       0.153
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]    -0.1013      0.027     -3.796      0.000      -0.154      -0.049
C(educacao, Treatment("Secundário"))[T.Superior completo]      0.0652      0.015      4.328      0.000       0.036       0.095
C(estado_civil, Treatment("Solteiro"))[T.Casado]              -0.1869      0.121     -1.551      0.121      -0.423       0.049
C(estado_civil, Treatment("Solteiro"))[T.União]               -0.2763      0.123     -2.248      0.025      -0.517      -0.035
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]               -0.0825      0.048     -1.711      0.087      -0.177       0.012
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]               0.1898      0.060      3.166      0.002       0.072       0.307
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]           0.0535      0.033      1.641      0.101      -0.010       0.117
C(tipo_residencia, Treatment("Casa"))[T.Comunitário]          -0.1204      0.101     -1.196      0.232      -0.318       0.077
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]               0.3186      0.093      3.424      0.001       0.136       0.501
posse_de_veiculo[T.True]                                       0.2815      0.015     18.894      0.000       0.252       0.311
posse_de_imovel[T.True]                                        0.0596      0.015      3.878      0.000       0.029       0.090
qtd_filhos                                                    -0.2050      0.120     -1.712      0.087      -0.440       0.030
idade                                                          0.0028      0.001      3.171      0.002       0.001       0.004
tempo_emprego                                                  0.0571      0.001     49.700      0.000       0.055       0.059
qt_pessoas_residencia                                          0.2412      0.120      2.017      0.044       0.007       0.476
==============================================================================
Omnibus:                       24.505   Durbin-Watson:                   2.033
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.593
Skew:                           0.109   Prob(JB):                     4.57e-06
Kurtosis:                       3.018   Cond. No.                     1.59e+03
=====================

In [99]:
x = x.drop('C(tipo_residencia, Treatment("Casa"))[T.Comunitário]', axis=1)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.207
Method:                 Least Squares   F-statistic:                     203.2
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:10:37   Log-Likelihood:                -14870.
No. Observations:               12427   AIC:                         2.977e+04
Df Residuals:                   12410   BIC:                         2.990e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================================================================
                                                                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      7.3092      0.127     57.583      0.000       7.060       7.558
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]         -0.1123      0.017     -6.798      0.000      -0.145      -0.080
C(tipo_renda, Treatment("Empresário"))[T.Pensionista]         -0.3716      0.268     -1.388      0.165      -0.896       0.153
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]    -0.1018      0.027     -3.815      0.000      -0.154      -0.050
C(educacao, Treatment("Secundário"))[T.Superior completo]      0.0643      0.015      4.274      0.000       0.035       0.094
C(estado_civil, Treatment("Solteiro"))[T.Casado]              -0.1869      0.121     -1.551      0.121      -0.423       0.049
C(estado_civil, Treatment("Solteiro"))[T.União]               -0.2757      0.123     -2.243      0.025      -0.517      -0.035
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]               -0.0838      0.048     -1.737      0.082      -0.178       0.011
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]               0.1906      0.060      3.178      0.001       0.073       0.308
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]           0.0543      0.033      1.667      0.096      -0.010       0.118
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]               0.3194      0.093      3.433      0.001       0.137       0.502
posse_de_veiculo[T.True]                                       0.2811      0.015     18.874      0.000       0.252       0.310
posse_de_imovel[T.True]                                        0.0599      0.015      3.902      0.000       0.030       0.090
qtd_filhos                                                    -0.2048      0.120     -1.710      0.087      -0.440       0.030
idade                                                          0.0028      0.001      3.160      0.002       0.001       0.004
tempo_emprego                                                  0.0571      0.001     49.762      0.000       0.055       0.059
qt_pessoas_residencia                                          0.2411      0.120      2.016      0.044       0.007       0.475
==============================================================================
Omnibus:                       24.456   Durbin-Watson:                   2.033
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.541
Skew:                           0.108   Prob(JB):                     4.69e-06
Kurtosis:                       3.020   Cond. No.                     1.59e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the 

In [101]:
x = x.drop('C(tipo_renda, Treatment("Empresário"))[T.Pensionista]', axis=1)

sm.OLS(y, x).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.207
Method:                 Least Squares   F-statistic:                     216.6
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:11:59   Log-Likelihood:                -14871.
No. Observations:               12427   AIC:                         2.977e+04
Df Residuals:                   12411   BIC:                         2.989e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================================================================
                                                                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      7.3069      0.127     57.567      0.000       7.058       7.556
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]         -0.1114      0.017     -6.750      0.000      -0.144      -0.079
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]    -0.1008      0.027     -3.780      0.000      -0.153      -0.049
C(educacao, Treatment("Secundário"))[T.Superior completo]      0.0638      0.015      4.243      0.000       0.034       0.093
C(estado_civil, Treatment("Solteiro"))[T.Casado]              -0.1859      0.121     -1.542      0.123      -0.422       0.050
C(estado_civil, Treatment("Solteiro"))[T.União]               -0.2746      0.123     -2.234      0.026      -0.515      -0.034
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]               -0.0829      0.048     -1.720      0.085      -0.177       0.012
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]               0.1911      0.060      3.187      0.001       0.074       0.309
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]           0.0548      0.033      1.682      0.093      -0.009       0.119
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]               0.3198      0.093      3.437      0.001       0.137       0.502
posse_de_veiculo[T.True]                                       0.2811      0.015     18.869      0.000       0.252       0.310
posse_de_imovel[T.True]                                        0.0602      0.015      3.920      0.000       0.030       0.090
qtd_filhos                                                    -0.2049      0.120     -1.711      0.087      -0.440       0.030
idade                                                          0.0028      0.001      3.178      0.001       0.001       0.004
tempo_emprego                                                  0.0571      0.001     49.753      0.000       0.055       0.059
qt_pessoas_residencia                                          0.2411      0.120      2.016      0.044       0.007       0.476
==============================================================================
Omnibus:                       24.758   Durbin-Watson:                   2.033
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.847
Skew:                           0.109   Prob(JB):                     4.02e-06
Kurtosis:                       3.019   Cond. No.                     1.59e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.59e+03. This might indicate that there are
strong multicoll

In [103]:
x = x.drop('C(estado_civil, Treatment("Solteiro"))[T.Casado]', axis=1)

sm.OLS(y, x).fit().summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.207
Model:                            OLS   Adj. R-squared:                  0.206
Method:                 Least Squares   F-statistic:                     231.9
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:12:45   Log-Likelihood:                -14872.
No. Observations:               12427   AIC:                         2.977e+04
Df Residuals:                   12412   BIC:                         2.989e+04
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
==============================================================================================================================
                                                                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      7.4864      0.051    148.042      0.000       7.387       7.586
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]         -0.1113      0.017     -6.740      0.000      -0.144      -0.079
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]    -0.1005      0.027     -3.768      0.000      -0.153      -0.048
C(educacao, Treatment("Secundário"))[T.Superior completo]      0.0637      0.015      4.235      0.000       0.034       0.093
C(estado_civil, Treatment("Solteiro"))[T.União]               -0.0900      0.028     -3.224      0.001      -0.145      -0.035
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]               -0.0802      0.048     -1.665      0.096      -0.175       0.014
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]               0.1916      0.060      3.196      0.001       0.074       0.309
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]           0.0552      0.033      1.694      0.090      -0.009       0.119
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]               0.3201      0.093      3.440      0.001       0.138       0.503
posse_de_veiculo[T.True]                                       0.2805      0.015     18.833      0.000       0.251       0.310
posse_de_imovel[T.True]                                        0.0604      0.015      3.935      0.000       0.030       0.091
qtd_filhos                                                    -0.0235      0.022     -1.045      0.296      -0.068       0.021
idade                                                          0.0028      0.001      3.172      0.002       0.001       0.004
tempo_emprego                                                  0.0571      0.001     49.751      0.000       0.055       0.059
qt_pessoas_residencia                                          0.0591      0.019      3.094      0.002       0.022       0.096
==============================================================================
Omnibus:                       24.493   Durbin-Watson:                   2.034
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.579
Skew:                           0.108   Prob(JB):                     4.60e-06
Kurtosis:                       3.020   Cond. No.                         552.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [105]:
x = x.drop('qtd_filhos', axis=1)

sm.OLS(y, x).fit().summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.207
Model:                            OLS   Adj. R-squared:                  0.206
Method:                 Least Squares   F-statistic:                     249.7
Date:                Tue, 28 Jan 2025   Prob (F-statistic):               0.00
Time:                        17:13:58   Log-Likelihood:                -14873.
No. Observations:               12427   AIC:                         2.977e+04
Df Residuals:                   12413   BIC:                         2.988e+04
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================================================================
                                                                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      7.5099      0.045    165.814      0.000       7.421       7.599
C(tipo_renda, Treatment("Empresário"))[T.Assalariado]         -0.1111      0.017     -6.728      0.000      -0.143      -0.079
C(tipo_renda, Treatment("Empresário"))[T.Servidor público]    -0.1004      0.027     -3.762      0.000      -0.153      -0.048
C(educacao, Treatment("Secundário"))[T.Superior completo]      0.0633      0.015      4.206      0.000       0.034       0.093
C(estado_civil, Treatment("Solteiro"))[T.União]               -0.0857      0.028     -3.104      0.002      -0.140      -0.032
C(estado_civil, Treatment("Solteiro"))[T.Viúvo]               -0.0937      0.046     -2.017      0.044      -0.185      -0.003
C(tipo_residencia, Treatment("Casa"))[T.Aluguel]               0.1902      0.060      3.173      0.002       0.073       0.308
C(tipo_residencia, Treatment("Casa"))[T.Com os pais]           0.0544      0.033      1.669      0.095      -0.009       0.118
C(tipo_residencia, Treatment("Casa"))[T.Estúdio]               0.3192      0.093      3.431      0.001       0.137       0.502
posse_de_veiculo[T.True]                                       0.2820      0.015     19.024      0.000       0.253       0.311
posse_de_imovel[T.True]                                        0.0601      0.015      3.917      0.000       0.030       0.090
idade                                                          0.0029      0.001      3.389      0.001       0.001       0.005
tempo_emprego                                                  0.0571      0.001     49.745      0.000       0.055       0.059
qt_pessoas_residencia                                          0.0410      0.008      5.058      0.000       0.025       0.057
==============================================================================
Omnibus:                       24.734   Durbin-Watson:                   2.033
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.823
Skew:                           0.109   Prob(JB):                     4.07e-06
Kurtosis:                       3.019   Cond. No.                         552.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

-Modelo inicial: R-squared = 0.208, Adj. R-squared = 0.207.

-Modelo final: R-squared = 0.207, Adj. R-squared = 0.206.

A capacidade explicativa do modelo final é praticamente igual à do modelo inicial, indicando que as variáveis removidas tinham pouca ou nenhuma contribuição explicativa.


-Inicial: AIC = 29780.

-Final: AIC = 29770.

Ambos o critério melhororou levemente no modelo final, indicando maior eficiência do modelo em relação a troca entre ajuste e número de variáveis.